### PR TITLE
[b/547849846] Fix custom fields pagination in MP tool

### PR DESCRIPTION
**Pull request by adarshtwy**
- Implement _paginate_1p_get in BackendAPI to handle pageToken iteration across paginated 1P endpoints with cyclical token detection, root_response_key parameter, and safe null-item handling.
- Update list_installed_integrations, list_integration_instances, list_custom_fields, and list_alert_grouping_rules to paginate results.
- Add comprehensive unit tests covering multi-page token traversal, cyclical token guard, null items, 204 No Content, custom root keys, and JSON decode errors.
- Fixes issue where list_custom_fields truncated results to the first 50 items.

### CODE REVIEW

File: packages/mp/src/mp/dev_env/api.py [hunk 1]
```diff
@@ -178,6 +179,53 @@ def download_integration(self, integration_name: str) -> Response:
...
+                if not page_token or not items or page_token in seen_tokens:
+                    break
+                seen_tokens.add(page_token)
+            elif isinstance(data, list):
+                all_items.extend(data)
+                break
+            else:
+                break
+
+        return all_items
```
Comment (KrishnaSharma06) [inline packages/mp/src/mp/dev_env/api.py]: We can more enhance it with following updates.

---

### Why This Version Is Much Better

1. **Memory Optimization via Streaming (`Iterator[dict[str, Any]]`)**:
   - By implementing `_paginate_1p_get_stream` as a generator (`yield item`), it processes and streams items one-by-one with $O(1)$ memory consumption instead of buffering potentially tens of thousands of records in memory ($O(N)$).

2. **Layered Architecture (Streaming Core + Eager Convenience Wrapper)**:
   - Providing both `_paginate_1p_get_stream` (for lazy evaluation/large datasets) and `_paginate_1p_get` (as a thin `list(...)` wrapper for simpler use cases) gives callers full flexibility without duplicating pagination logic.

3. **Configurable Token Keys**:
   - `token_param_key: str = "pageToken"` and `token_response_key: str = "nextPageToken"` make this helper adaptable across different 1P/GCP APIs (e.g., services that use `next_page_token`, `pageToken`, or `nextPageToken`).

4. **Circuit Breaker Protection (`max_pages`)**:
   - Adding `max_pages: int | None = None` provides an essential circuit breaker to safely fetch bounded subsets of pages or sample data without runaway API calls.

5. **Resilience to Empty Filtered Pages**:
   - The loop checks `if not page_token: break` rather than breaking on `not items`. This ensures that if a 1P API returns an empty page (`items: []`) due to server-side filtering but includes a valid `nextPageToken`, pagination continues correctly.

6. **Defensive Typing & Param Isolation**:
   - `dict(params) if params else {}` prevents mutating the caller's dictionary.
   - `if isinstance(item, dict): yield item` prevents malformed non-dictionary elements from reaching downstream consumers.

---

```
    def _paginate_1p_get_stream(
        self,
        url: str,
        params: dict[str, Any] | None = None,
        root_response_key: str = "items",
        page_size: int | None = None,
        token_param_key: str = "pageToken",
        token_response_key: str = "nextPageToken",
        max_pages: int | None = None,
    ) -> Iterator[dict[str, Any]]:
        """Yield items across paginated requests without buffering entire datasets in memory.
        Args:
            url: The 1P endpoint URL.
            params: Optional query parameters for the request.
            root_response_key: The response JSON key containing the item list. Defaults to "items".
            page_size: Optional page size limit to request per page.
            token_param_key: Query parameter key used to pass the token. Defaults to "pageToken".
            token_response_key: Response key containing the next page token. Defaults to "nextPageToken".
            max_pages: Optional maximum number of pages to fetch before stopping.
        Yields:
            dict[str, Any]: Individual item records from each page.
        Raises:
            requests.HTTPError: If an HTTP request returns an error status code.
            requests.exceptions.JSONDecodeError: If a response body cannot be decoded as JSON.
        """
        page_token: str | None = None
        seen_tokens: set[str] = set()
        pages_fetched = 0
        request_params: dict[str, Any] = dict(params) if params else {}
        if page_size is not None:
            request_params["pageSize"] = page_size
        while True:
            if max_pages is not None and pages_fetched >= max_pages:
                logger.debug("Reached max page limit of %d for URL: %s", max_pages, url)
                break
            if page_token:
                request_params[token_param_key] = page_token
            else:
                request_params.pop(token_param_key, None)
            resp = self.session.get(url, params=request_params, timeout=DEFAULT_REQUEST_TIMEOUT)
            if not resp.ok:
                logger.error(
                    "Request to '%s' failed: %d - %s",
                    url,
                    resp.status_code,
                    resp.text,
                )
                resp.raise_for_status()
            payload = resp.json()
            if not isinstance(payload, dict):
                logger.error("Expected dictionary response from '%s', got %s", url, type(payload))
                break
            items = payload.get(root_response_key, [])
            if isinstance(items, list):
                for item in items:
                    if isinstance(item, dict):
                        yield item
            page_token = payload.get(token_response_key)
            pages_fetched += 1
            if not page_token:
                break
            if page_token in seen_tokens:
                logger.warning(
                    "Detected duplicate page token '%s' from URL '%s'. Breaking to prevent loop.",
                    page_token,
                    url,
                )
                break
            seen_tokens.add(page_token)
    def _paginate_1p_get(
        self,
        url: str,
        params: dict[str, Any] | None = None,
        root_response_key: str = "items",
        page_size: int | None = None,
        token_param_key: str = "pageToken",
        token_response_key: str = "nextPageToken",
        max_pages: int | None = None,
    ) -> list[dict[str, Any]]:
        """Fetch all items from a 1P API endpoint with pagination support.
        Args:
            url: The 1P endpoint URL.
            params: Optional query parameters for the request.
            root_response_key: The response JSON key containing the item list. Defaults to "items".
            page_size: Optional page size limit to request per page.
            token_param_key: Query parameter key used to pass the token. Defaults to "pageToken".
            token_response_key: Response key containing the next page token. Defaults to "nextPageToken".
            max_pages: Optional maximum number of pages to fetch before stopping.
        Returns:
            list[dict[str, Any]]: Aggregated list of items from all pages.
        """
        return list(
            self._paginate_1p_get_stream(
                url=url,
                params=params,
                root_response_key=root_response_key,
                page_size=page_size,
                token_param_key=token_param_key,
                token_response_key=token_response_key,
                max_pages=max_pages,
            )
        )
```

File: packages/mp/src/mp/dev_env/api.py [hunk 2]
```diff
@@ -186,14 +309,7 @@ def list_installed_integrations(self) -> list[dict[str, Any]]:
...
         url: str = f"{self.api_root}/api/1p/external/v1/integrations"
-        resp = self.session.get(url)
-        resp.raise_for_status()
-        if resp.status_code == 204:  # ruff:ignore[magic-value-comparison]
-            return []
-        data = resp.json()
-        if isinstance(data, dict):
-            return data.get("items", [])
-        return data if isinstance(data, list) else []
+        return self._paginate_1p_get(url)
```
Comment (KrishnaSharma06) [inline packages/mp/src/mp/dev_env/api.py]: I think we should not dependent on "items" key, so it is better to explicitly pass the root_response_key argument while calling for each API to get and paginate the correct response.

**Review by KrishnaSharma06 — CHANGES_REQUESTED** [review]

File: packages/mp/src/mp/dev_env/api.py (hunk 2 above)
Comment (adarshtwy) [reply packages/mp/src/mp/dev_env/api.py]: Fixed

**Review by KrishnaSharma06 — APPROVED** [review]

**Review by patilgp-oss — APPROVED** [review]

**Review by nitishgangal-ind — APPROVED** [review]

**Review by harishkardam — APPROVED** [review]

**Review by KrishnaSharma06 — APPROVED** [review]

**Review by AmitJ98 — APPROVED** [review]
